pandas-stubs/core/indexes/base.pyi

```diff
@@ -48,6 +50,22 @@ class InvalidIndexError(Exception): ...
 
 _str = str
 
+class _IndexGetitemMixin(Generic[S1]):
+    # type ignore needed because it doesn't like the type of self
+    @overload
+    def __getitem__(  # type: ignore[misc]
+        self: IndexT,
```

Comment:
I assume `only` Index inherits from this. Can you use `_IndexGetitemMixinT` for self/return value?

Reply:
Could probably also start using `typing.Self`

Reply:
> I assume `only` Index inherits from this. Can you use `_IndexGetitemMixinT` for self/return value?

Your assumption isn't correct.  `DatetimeIndex`, `PeriodIndex` and `TimedeltaIndex` also inherit, and then are specifying the return type of `__getitem__(x: int)` via the `Generic` parameter.  

> Could probably also start using typing.Self

I tried creating a `TypeVar` as  `_IndexGetitemMixinT` and it did work.  Tried `typing.Self`, and it doesn't work.  I think the type of `self` in this case has to be "bound" to the return type.

These changes are in commit https://github.com/pandas-dev/pandas-stubs/pull/537/commits/0db8a00a7bd5c7f023991459d79f372df99b1ab3